### DISCUSSION TITLE
Why Do We Need to Authenticate to Download/Read Ruby Gems?

### DISCUSSION THREAD

**Discussion by mzagaja** [community/community#38681] [Actions]
In taking GitHub Packages for a test drive, I found that the provided instructions/documentation on the Package page did not work. The reason is that it asks you to run:

```
gem install doc_to_dash --version "0.0.11" --source "https://rubygems.pkg.github.com/mzagaja"
```

But what you actually need to run is something like:

```
gem install doc_to_dash --version "0.0.11" --source "https://USERNAME:TOKEN@rubygems.pkg.github.com/mzagaja"
```

This creates a three part barrier to adopting GitHub Packages by end users:
1. They need to surmount the misdirection from the documentation and error it causes.
2. They need to have a GitHub account.
3. They need to learn about and generate a Personal Access Token (which oddly warns you it is **legacy** when you do it).

Using gems from rubygems.org does not trigger any of these issues.

**Comment by ossanna16** [comment]
Hi there @mzagaja and welcome to our community! Thank you for asking a great question 🙂
While you are waiting for a reply to your question, here are a few things you can do:

To get started, introduce yourself in our [official introduction thread](https://github.com/orgs/community/discussions/23204)
To connect with other new users, check out this Discussion to [Meet Your GitHub Community Team](https://github.com/orgs/community/discussions/28776), further your GitHub knowledge by working through our [GitHub Skill 🆙 exercises](https://github.com/orgs/community/discussions/32121), and participate in our [weekly Fun Friday chats](https://github.com/orgs/community/discussions/30912)

**Comment by spenserblack** [comment]
> But what you actually need to run is something like:
> 
> ```
> gem install doc_to_dash --version "0.0.11" --source "https://USERNAME:TOKEN@rubygems.pkg.github.com/mzagaja"
> ```

Just a clarification that [GitHub suggests putting the token in a `.gemrc` file](https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-rubygems-registry#installing-a-package).

But I also didn't expect this behavior. It makes sense for private repositories, but why should the package require authentication in public repos? Seems to me one would want the package to be available if they've made the repo public.

**Reply by mzagaja** [reply]
I agree, it makes sense to authenticate private repos, but not for public ones. Another thing I've come to understand is there is a way to universally authenticate with bundler and GitHub by setting:
```
export BUNDLE_GITHUB__COM = username:personal_access_token
```

But I have not seen a similar option documented for Rubygems which is a bit of a bummer and barrier. This means for every GitHub package you want to access for Rubygems without bundler, you need to add the source with your credentials to `.gemrc` or `gem sources --add`.

 I will also note using `.gemrc` somewhat defeats the suggested installation procedure. With my private access token and that URI in my .gemrc as suggested I get:
```
gem install doc_to_dash --version "0.0.11" --source "https://rubygems.pkg.github.com/mzagaja"
Successfully installed doc_to_dash-0.0.11
Parsing documentation for doc_to_dash-0.0.11
Installing ri documentation for doc_to_dash-0.0.11
Done installing documentation for doc_to_dash after 0 seconds
WARNING:  Unable to pull data from 'https://rubygems.pkg.github.com/mzagaja/': bad response Unauthorized 401 (https://rubygems.pkg.github.com/mzagaja/specs.4.8.gz)
1 gem installed
```
I suspect that the only reason this succeeded is that the gem was previously cached somewhere on this machine versus actually successfully mapping the `.gemrc` credential URI to the one I pass to `--source` but am not certain.

**Reply by spenserblack** [reply]
I just quickly went through the docs again and it looks like all package registries, with the exception of the container registry, require authentication to install/pull/download (whatever terminology for the given package type). Perhaps the intended usage was for private packages for organizations, but, frankly, my intended usage is to publish packages without needing to create an account with yet another package repository :laughing:
